src/diagram/join.rs

```diff
@@ -0,0 +1,314 @@
+use std::any::TypeId;
+
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+use smallvec::SmallVec;
+use tracing::debug;
+
+use crate::{Builder, IterBufferable, Output};
+
+use super::{
+    DiagramError, DynOutput, NextOperation, NodeRegistry, SerializeMessage, SourceOperation,
+};
+
+#[derive(Debug, Serialize, Deserialize, JsonSchema)]
+#[serde(rename_all = "snake_case")]
+pub struct JoinOp {
+    pub(super) next: NextOperation,
+
+    /// Controls the order of the resulting join. Each item must be an operation id of one of the
+    /// incoming outputs.
+    pub(super) order: Vec<SourceOperation>,
```

Comment:
I understand why you named this `order`, but it will make more sense for it to be named `inputs` or `sources` once we add support for buffers. Buffers won't have a concept of `next`, so this field will be the only place where we will see a connection between a `buffer` and a `join`. At that point it's describing more than just an order; it's describing what its inputs are.